File: js/blackscholes.js
/**
 * Eagle System — Black-Scholes Engine
 * Implementação correta do modelo Black-Scholes para:
 *   - Preço teórico de opções europeias (Call e Put)
 *   - Gregas: Delta, Gamma, Theta, Vega, Rho
 *   - Volatilidade Implícita (Newton-Raphson)
 *   - Cálculo de payoff na expiração e antes (usando BS)
 */

'use strict';

const BS = (() => {

    // Função de distribuição normal acumulada (CDF) — Abramowitz & Stegun
    function normCDF(x) {
        const a1 = 0.254829592;
        const a2 = -0.284496736;
        const a3 = 1.421413741;
        const a4 = -1.453152027;
        const a5 = 1.061405429;
        const p = 0.3275911;
        const sign = x < 0 ? -1 : 1;
        x = Math.abs(x) / Math.sqrt(2);
        const t = 1.0 / (1.0 + p * x);
        const y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);
        return 0.5 * (1.0 + sign * y);
    }

    // Função densidade de probabilidade normal (PDF)
    function normPDF(x) {
        return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
    }

    /**
     * Calcula d1 e d2 do modelo Black-Scholes
     * @param {number} S  - Preço spot do ativo
     * @param {number} K  - Strike da opção
     * @param {number} T  - Tempo até vencimento em ANOS
     * @param {number} r  - Taxa de juros livre de risco (ex: 0.10 para 10%)
     * @param {number} sigma - Volatilidade implícita anualizada (ex: 0.30 para 30%)
     * @param {number} [q=0] - Dividend yield contínuo
     */
    function d1d2(S, K, T, r, sigma, q = 0) {
        if (T <= 0 || sigma <= 0 || S <= 0 || K <= 0) return { d1: NaN, d2: NaN };
        const d1 = (Math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
        const d2 = d1 - sigma * Math.sqrt(T);
        return { d1, d2 };
    }

    /**
     * Preço Black-Scholes de uma opção europeia
     * @param {'call'|'put'} type
     * @param {number} S  - Preço spot
     * @param {number} K  - Strike
     * @param {number} T  - Tempo em anos
     * @param {number} r  - Taxa livre de risco
     * @param {number} sigma - Volatilidade
     * @param {number} [q=0] - Dividend yield
     * @returns {number} Preço teórico
     */
    function price(type, S, K, T, r, sigma, q = 0) {
        if (T <= 0) {
            // No vencimento: valor intrínseco
            if (type === 'call') return Math.max(0, S - K);
            return Math.max(0, K - S);
        }
        const { d1, d2 } = d1d2(S, K, T, r, sigma, q);
        if (isNaN(d1)) return 0;
        if (type === 'call') {
            return S * Math.exp(-q * T) * normCDF(d1) - K * Math.exp(-r * T) * normCDF(d2);
        } else {
            return K * Math.exp(-r * T) * normCDF(-d2) - S * Math.exp(-q * T) * normCDF(-d1);
        }
    }

    /**
     * Gregas Black-Scholes completas
     * @returns {{ delta, gamma, theta, vega, rho }}
     *   theta: por DIA (dividido por 365)
     *   vega:  por 1% de mudança em vol (dividido por 100)
     *   rho:   por 1% de mudança na taxa (dividido por 100)
     */
    function greeks(type, S, K, T, r, sigma, q = 0) {
        if (T <= 0 || sigma <= 0 || S <= 0 || K <= 0) {
            return { delta: null, gamma: null, theta: null, vega: null, rho: null };
        }
        const { d1, d2 } = d1d2(S, K, T, r, sigma, q);
        const Nd1 = normCDF(d1);
        const Nd2 = normCDF(d2);
        const nd1 = normPDF(d1);
        const eqT = Math.exp(-q * T);
        const erT = Math.exp(-r * T);
        const sqT = Math.sqrt(T);

        let delta, theta, rho;

        if (type === 'call') {
            delta = eqT * Nd1;
            theta = (-(S * eqT * nd1 * sigma) / (2 * sqT)
                - r * K * erT * Nd2
                + q * S * eqT * Nd1) / 365;
            rho = K * T * erT * Nd2 / 100;
        } else {
            delta = -eqT * normCDF(-d1);
            theta = (-(S * eqT * nd1 * sigma) / (2 * sqT)
                + r * K * erT * normCDF(-d2)
                - q * S * eqT * normCDF(-d1)) / 365;
            rho = -K * T * erT * normCDF(-d2) / 100;
        }

        const gamma = (eqT * nd1) / (S * sigma * sqT);
        const vega = S * eqT * nd1 * sqT / 100;

        return { delta, gamma, theta, vega, rho };
    }

    /**
     * Volatilidade Implícita via Newton-Raphson
     * @param {'call'|'put'} type
     * @param {number} S, K, T, r, q
     * @param {number} marketPrice - Preço de mercado da opção
     * @returns {number|null} IV (0.01 a 5.00) ou null se não convergir
     */
    function impliedVol(type, S, K, T, r, marketPrice, q = 0) {
        if (T <= 0 || marketPrice <= 0) return null;
        let sigma = 0.3; // chute inicial: 30%
        for (let i = 0; i < 100; i++) {
            const { d1 } = d1d2(S, K, T, r, sigma, q);
            const p = price(type, S, K, T, r, sigma, q);
            const vg = S * Math.exp(-q * T) * normPDF(d1) * Math.sqrt(T); // vega sem divisão por 100
            const diff = p - marketPrice;
            if (Math.abs(diff) < 1e-8) break;
            if (Math.abs(vg) < 1e-10) break;
            sigma = sigma - diff / vg;
            if (sigma <= 0) sigma = 1e-5;
            if (sigma > 10) return null;
        }
        return sigma >= 0.001 && sigma <= 10 ? sigma : null;
    }

    /**
     * Payoff de UMA perna na expiração (T=0)
     * Fórmula exata: valor intrínseco menos prêmio pago/recebido
     * @param {object} leg - { operacao:'compra'|'venda', tipo:'call'|'put', strike, premio, qtd }
     * @param {number} S   - Preço do ativo na expiração
     * @returns {number} P&L da perna no vencimento
     */
    function legPayoffAtExpiry(leg, S) {
        const K = parseFloat(leg.strike) || 0;
        const prem = parseFloat(leg.premio) || 0;
        const qtd = parseFloat(leg.qtd) || 1;
        const mult = leg.operacao === 'compra' ? 1 : -1; // compra: long, venda: short

        let intrinsic;
        if (leg.tipo === 'call') intrinsic = Math.max(0, S - K);
        else intrinsic = Math.max(0, K - S);

        // P&L = direção × qtd × (valor_intrínseco - prêmio_pago)
        // compra: pagamos prêmio → P&L = intrínseco - prêmio
        // venda:  recebemos prêmio → P&L = -(intrínseco - prêmio) = prêmio - intrínseco
        return mult * qtd * (intrinsic - prem);
    }

    /**
     * P&L atual de uma perna com Black-Scholes (antes do vencimento)
     * @param {object} leg  - com iv (volatilidade implícita em %)
     * @param {number} S    - Preço atual
     * @param {number} T    - Tempo até vencimento em anos
     * @param {number} r    - Taxa de juros
     * @returns {number} P&L atual da perna
     */
    function legPnLCurrent(leg, S, T, r = 0.1075) {
        const K = parseFloat(leg.strike) || 0;
        const prem = parseFloat(leg.premio) || 0;
        const qtd = parseFloat(leg.qtd) || 1;
        const iv = parseFloat(leg.iv) || 30;
        const mult = leg.operacao === 'compra' ? 1 : -1;
        const sigma = iv / 100;
        const currentValue = price(leg.tipo, S, K, T, r, sigma);
        return mult * qtd * (currentValue - prem);
    }

    /**
     * Gera série de pontos para o gráfico de payoff
     * @param {object} estrutura - { precoAtual, dataVenc }
     * @param {Array}  legs
     * @param {number} [numPoints=300]
     * @returns {{ labels, payoffExpiry, payoffCurrent, xMin, xMax }}
     */
    function computePayoffSeries(estrutura, legs, numPoints = 300) {
        if (!legs.length) return null;

        const S0 = parseFloat(estrutura.precoAtual) || 20;
        const r = 0.1075; // Selic aproximada

        // Tempo até vencimento
        let T = 0;
        if (estrutura.dataVenc) {
            const hoje = new Date();
            const venc = new Date(estrutura.dataVenc + 'T23:59:59');
            const dias = (venc - hoje) / (1000 * 60 * 60 * 24);
            T = Math.max(0, dias / 365);
        }

        // Range de preços: de 50% a 150% do spot, alargando para cobrir strikes
        const strikes = legs.map(l => parseFloat(l.strike)).filter(s => s > 0);
        const minStr = Math.min(S0, ...strikes);
        const maxStr = Math.max(S0, ...strikes);
        const xMin = Math.max(0.01, minStr * 0.6);
        const xMax = maxStr * 1.4;

        const labels = [];
        const payoffExpiry = [];
        const payoffCurrent = [];

        for (let i = 0; i <= numPoints; i++) {
            const S = xMin + (xMax - xMin) * (i / numPoints);
            labels.push(S);

            // Payoff na expiração (T=0, valor intrínseco)
            let pnlExp = 0;
            for (const leg of legs) {
                pnlExp += legPayoffAtExpiry(leg, S);
            }
            payoffExpiry.push(pnlExp);

            // P&L atual (Black-Scholes se T > 0)
            if (T > 0.001) {
                let pnlCur = 0;
                for (const leg of legs) {
                    pnlCur += legPnLCurrent(leg, S, T, r);
                }
                payoffCurrent.push(pnlCur);
            }
        }

        return { labels, payoffExpiry, payoffCurrent: payoffCurrent.length ? payoffCurrent : null, xMin, xMax, T, S0 };
    }

    /**
     * Calcula métricas completas da estrutura
     */
    function calcMetrics(estrutura, legs) {
        const zero = { posInicial: 0, ganhoMax: 0, perdaMax: 0, breakEvens: [], margem: 0, premioLiq: 0, delta: null, gamma: null, theta: null, vega: null, rho: null };
        if (!legs.length) return zero;

        const S0 = parseFloat(estrutura.precoAtual) || 0;
        const r = 0.1075;

        // Tempo até vencimento
        let T = 0;
        if (estrutura.dataVenc) {
            const hoje = new Date();
            const venc = new Date(estrutura.dataVenc + 'T23:59:59');
            const dias = (venc - hoje) / (1000 * 60 * 60 * 24);
            T = Math.max(0, dias / 365);
        }

        // Prêmio líquido: compra paga, venda recebe
        let premioLiq = 0;
        let gDelta = 0, gGamma = 0, gTheta = 0, gVega = 0, gRho = 0;
        let hasGreeks = false;

        for (const leg of legs) {
            const prem = parseFloat(leg.premio) || 0;
            const qtd = parseFloat(leg.qtd) || 1;
            const mult = leg.operacao === 'compra' ? -1 : 1; // compra paga, venda recebe
            premioLiq += mult * prem * qtd;

            // Gregas: usar BS se tiver IV, senão usar valores manuais
            const K = parseFloat(leg.strike) || 0;
            const iv_pct = parseFloat(leg.iv);
            const sigma = (iv_pct > 0) ? iv_pct / 100 : null;

            let g;
            if (sigma && T > 0 && S0 > 0 && K > 0) {
                g = greeks(leg.tipo, S0, K, T, r, sigma);
                hasGreeks = true;
            } else if (leg.delta !== null && leg.delta !== '') {
                // Usar valores manuais
                g = {
                    delta: parseFloat(leg.delta) || 0,
                    gamma: parseFloat(leg.gamma) || 0,
                    theta: parseFloat(leg.theta) || 0,
                    vega: parseFloat(leg.vega) || 0,
                    rho: 0
                };
                hasGreeks = true;
            } else {
                g = null;
            }

            if (g) {
                const d = leg.operacao === 'compra' ? 1 : -1;
                gDelta += d * (g.delta || 0) * qtd;
                gGamma += d * (g.gamma || 0) * qtd;
                gTheta += d * (g.theta || 0) * qtd;
                gVega += d * (g.vega || 0) * qtd;
                gRho += d * (g.rho || 0) * qtd;
            }
        }

        // Calcular payoff para encontrar max/min/breakeven
        const series = computePayoffSeries(estrutura, legs, 500);
        let ganhoMax = -Infinity, perdaMax = Infinity;
        const breakEvens = [];

        if (series) {
            for (const v of series.payoffExpiry) {
                if (v > ganhoMax) ganhoMax = v;
                if (v < perdaMax) perdaMax = v;
            }
            // Break-even: cruzamentos de zero
            for (let i = 0; i < series.payoffExpiry.length - 1; i++) {
                const y0 = series.payoffExpiry[i];
                const y1 = series.payoffExpiry[i + 1];
                if ((y0 <= 0 && y1 > 0) || (y0 >= 0 && y1 < 0)) {
                    const x0 = series.labels[i];
                    const x1 = series.labels[i + 1];
                    const xBE = x0 + (x1 - x0) * (-y0 / (y1 - y0));
                    breakEvens.push(xBE);
                }
            }
        }

        if (ganhoMax === -Infinity) ganhoMax = 0;
        if (perdaMax === Infinity) perdaMax = 0;

        // Margem: aproximação para estruturas de crédito (máxima perda)
        const margem = Math.abs(perdaMax) > 0 ? Math.abs(perdaMax) : 0;

        return {
            posInicial: premioLiq,
            ganhoMax,
            perdaMax,
            breakEvens,
            margem,
            premioLiq,
            delta: hasGreeks ? gDelta : null,
            gamma: hasGreeks ? gGamma : null,
            theta: hasGreeks ? gTheta : null,
            vega: hasGreeks ? gVega : null,
            rho: hasGreeks ? gRho : null,
        };
    }

    return { price, greeks, impliedVol, legPayoffAtExpiry, legPnLCurrent, computePayoffSeries, calcMetrics, normCDF };

})();

// Exporta para uso nos módulos
if (typeof module !== 'undefined') module.exports = BS;

File: system/core/blackscholes.py
"""
System — Black-Scholes Engine (port fiel do js/blackscholes.js).
"""

from __future__ import annotations

import math
from typing import Any, Literal, Optional, TypedDict

OptionType = Literal["call", "put"]


def norm_cdf(x: float) -> float:
    a1 = 0.254829592
    a2 = -0.284496736
    a3 = 1.421413741
    a4 = -1.453152027
    a5 = 1.061405429
    p = 0.3275911
    sign = -1.0 if x < 0 else 1.0
    ax = abs(x) / math.sqrt(2.0)
    t = 1.0 / (1.0 + p * ax)
    y = 1.0 - (((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * math.exp(-ax * ax))
    return 0.5 * (1.0 + sign * y)


def norm_pdf(x: float) -> float:
    return math.exp(-0.5 * x * x) / math.sqrt(2 * math.pi)


def d1d2(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0):
    if T <= 0 or sigma <= 0 or S <= 0 or K <= 0:
        return float("nan"), float("nan")
    d1 = (math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * math.sqrt(T))
    d2 = d1 - sigma * math.sqrt(T)
    return d1, d2


def price(
    typ: OptionType,
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    q: float = 0.0,
) -> float:
    if T <= 0:
        if typ == "call":
            return max(0.0, S - K)
        return max(0.0, K - S)
    d1, d2 = d1d2(S, K, T, r, sigma, q)
    if math.isnan(d1):
        return 0.0
    if typ == "call":
        return S * math.exp(-q * T) * norm_cdf(d1) - K * math.exp(-r * T) * norm_cdf(d2)
    return K * math.exp(-r * T) * norm_cdf(-d2) - S * math.exp(-q * T) * norm_cdf(-d1)


class GreeksResult(TypedDict, total=False):
    delta: Optional[float]
    gamma: Optional[float]
    theta: Optional[float]
    vega: Optional[float]
    rho: Optional[float]


def greeks(
    typ: OptionType,
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    q: float = 0.0,
) -> GreeksResult:
    if T <= 0 or sigma <= 0 or S <= 0 or K <= 0:
        return {
            "delta": None,
            "gamma": None,
            "theta": None,
            "vega": None,
            "rho": None,
        }
    d1, d2 = d1d2(S, K, T, r, sigma, q)
    nd1 = norm_pdf(d1)
    eq_t = math.exp(-q * T)
    er_t = math.exp(-r * T)
    sq_t = math.sqrt(T)
    nd1_cdf = norm_cdf(d1)
    nd2_cdf = norm_cdf(d2)

    if typ == "call":
        delta = eq_t * nd1_cdf
        theta = (
            -(S * eq_t * nd1 * sigma) / (2 * sq_t)
            - r * K * er_t * nd2_cdf
            + q * S * eq_t * nd1_cdf
        ) / 365.0
        rho = K * T * er_t * nd2_cdf / 100.0
    else:
        delta = -eq_t * norm_cdf(-d1)
        theta = (
            -(S * eq_t * nd1 * sigma) / (2 * sq_t)
            + r * K * er_t * norm_cdf(-d2)
            - q * S * eq_t * norm_cdf(-d1)
        ) / 365.0
        rho = -K * T * er_t * norm_cdf(-d2) / 100.0

    gamma = (eq_t * nd1) / (S * sigma * sq_t)
    vega = S * eq_t * nd1 * sq_t / 100.0

    return {"delta": delta, "gamma": gamma, "theta": theta, "vega": vega, "rho": rho}


def implied_vol(
    typ: OptionType,
    S: float,
    K: float,
    T: float,
    r: float,
    market_price: float,
    q: float = 0.0,
) -> Optional[float]:
    if T <= 0 or market_price <= 0:
        return None
    sigma = 0.3
    for _ in range(100):
        d1, _ = d1d2(S, K, T, r, sigma, q)
        p = price(typ, S, K, T, r, sigma, q)
        vg = S * math.exp(-q * T) * norm_pdf(d1) * math.sqrt(T)
        diff = p - market_price
        if abs(diff) < 1e-8:
            break
        if abs(vg) < 1e-10:
            break
        sigma = sigma - diff / vg
        if sigma <= 0:
            sigma = 1e-5
        if sigma > 10:
            return None
    if 0.001 <= sigma <= 10:
        return sigma
    return None


def _float_leg(leg: dict, key: str, default: float = 0.0) -> float:
    v = leg.get(key)
    if v is None or v == "":
        return default
    try:
        return float(v)
    except (TypeError, ValueError):
        return default


def leg_payoff_at_expiry(leg: dict, S: float) -> float:
    K = _float_leg(leg, "strike")
    prem = _float_leg(leg, "premio")
    qtd = _float_leg(leg, "qtd", 1.0)
    mult = 1.0 if leg.get("operacao") == "compra" else -1.0
    if leg.get("tipo") == "call":
        intrinsic = max(0.0, S - K)
    else:
        intrinsic = max(0.0, K - S)
    return mult * qtd * (intrinsic - prem)


def leg_pnl_current(leg: dict, S: float, T: float, r: float = 0.1075) -> float:
    K = _float_leg(leg, "strike")
    prem = _float_leg(leg, "premio")
    qtd = _float_leg(leg, "qtd", 1.0)
    iv = _float_leg(leg, "iv", 30.0)
    mult = 1.0 if leg.get("operacao") == "compra" else -1.0
    sigma = iv / 100.0
    typ = leg.get("tipo") or "call"
    if typ not in ("call", "put"):
        typ = "call"
    current_value = price(typ, S, K, T, r, sigma)  # type: ignore[arg-type]
    return mult * qtd * (current_value - prem)


def _days_to_expiry(data_venc: Optional[str]) -> float:
    if not data_venc:
        return 0.0
    from datetime import date, datetime

    try:
        venc = datetime.strptime(data_venc[:10], "%Y-%m-%d").replace(
            hour=23, minute=59, second=59
        )
        hoje = datetime.now()
        dias = (venc - hoje).total_seconds() / (24 * 3600)
        return max(0.0, dias / 365.0)
    except (ValueError, TypeError):
        return 0.0


def compute_payoff_series(
    estrutura: dict,
    legs: list,
    num_points: int = 300,
):
    if not legs:
        return None

    S0 = float(estrutura.get("precoAtual") or estrutura.get("preco_atual") or 20)
    r = 0.1075

    data_venc = estrutura.get("dataVenc") or estrutura.get("data_venc")
    T = _days_to_expiry(data_venc)

    strikes = []
    for leg in legs:
        k = _float_leg(leg, "strike")
        if k > 0:
            strikes.append(k)
    if not strikes:
        min_str = S0
        max_str = S0
    else:
        min_str = min(S0, min(strikes))
        max_str = max(S0, max(strikes))
    x_min = max(0.01, min_str * 0.6)
    x_max = max_str * 1.4

    labels: list[float] = []
    payoff_expiry: list[float] = []
    payoff_current: list[float] = []

    for i in range(num_points + 1):
        sx = x_min + (x_max - x_min) * (i / num_points)
        labels.append(sx)
        pnl_exp = 0.0
        for leg in legs:
            pnl_exp += leg_payoff_at_expiry(leg, sx)
        payoff_expiry.append(pnl_exp)
        if T > 0.001:
            pnl_cur = 0.0
            for leg in legs:
                pnl_cur += leg_pnl_current(leg, sx, T, r)
            payoff_current.append(pnl_cur)

    payoff_cur_out = payoff_current if payoff_current else None
    return {
        "labels": labels,
        "payoffExpiry": payoff_expiry,
        "payoffCurrent": payoff_cur_out,
        "xMin": x_min,
        "xMax": x_max,
        "T": T,
        "S0": S0,
    }


def calc_metrics(estrutura: dict, legs: list) -> dict[str, Any]:
    zero = {
        "posInicial": 0.0,
        "ganhoMax": 0.0,
        "perdaMax": 0.0,
        "breakEvens": [],
        "margem": 0.0,
        "premioLiq": 0.0,
        "delta": None,
        "gamma": None,
        "theta": None,
        "vega": None,
        "rho": None,
    }
    if not legs:
        return zero

    S0 = float(estrutura.get("precoAtual") or estrutura.get("preco_atual") or 0)
    r = 0.1075
    data_venc = estrutura.get("dataVenc") or estrutura.get("data_venc")
    T = _days_to_expiry(data_venc)

    premio_liq = 0.0
    g_delta = g_gamma = g_theta = g_vega = g_rho = 0.0
    has_greeks = False

    for leg in legs:
        prem = _float_leg(leg, "premio")
        qtd = _float_leg(leg, "qtd", 1.0)
        mult = -1.0 if leg.get("operacao") == "compra" else 1.0
        premio_liq += mult * prem * qtd

        K = _float_leg(leg, "strike")
        iv_pct = leg.get("iv")
        try:
            iv_val = float(iv_pct) if iv_pct is not None and iv_pct != "" else None
        except (TypeError, ValueError):
            iv_val = None
        sigma = iv_val / 100.0 if iv_val and iv_val > 0 else None

        typ = leg.get("tipo") or "call"
        if typ not in ("call", "put"):
            typ = "call"

        g = None
        if sigma and T > 0 and S0 > 0 and K > 0:
            g = greeks(typ, S0, K, T, r, sigma)
            has_greeks = True
        elif leg.get("delta") is not None and leg.get("delta") != "":
            g = {
                "delta": float(leg.get("delta") or 0),
                "gamma": float(leg.get("gamma") or 0),
                "theta": float(leg.get("theta") or 0),
                "vega": float(leg.get("vega") or 0),
                "rho": 0.0,
            }
            has_greeks = True

        if g:
            d = 1.0 if leg.get("operacao") == "compra" else -1.0
            g_delta += d * (g.get("delta") or 0) * qtd
            g_gamma += d * (g.get("gamma") or 0) * qtd
            g_theta += d * (g.get("theta") or 0) * qtd
            g_vega += d * (g.get("vega") or 0) * qtd
            g_rho += d * (g.get("rho") or 0) * qtd

    series = compute_payoff_series(estrutura, legs, 500)
    ganho_max = float("-inf")
    perda_max = float("inf")
    break_evens: list[float] = []

    if series:
        for v in series["payoffExpiry"]:
            if v > ganho_max:
                ganho_max = v
            if v < perda_max:
                perda_max = v
        pe = series["payoffExpiry"]
        lbl = series["labels"]
        for i in range(len(pe) - 1):
            y0, y1 = pe[i], pe[i + 1]
            if (y0 <= 0 and y1 > 0) or (y0 >= 0 and y1 < 0):
                x0, x1 = lbl[i], lbl[i + 1]
                x_be = x0 + (x1 - x0) * (-y0 / (y1 - y0)) if y1 != y0 else x0
                break_evens.append(x_be)

    if ganho_max == float("-inf"):
        ganho_max = 0.0
    if perda_max == float("inf"):
        perda_max = 0.0

    margem = abs(perda_max) if abs(perda_max) > 0 else 0.0

    return {
        "posInicial": premio_liq,
        "ganhoMax": ganho_max,
        "perdaMax": perda_max,
        "breakEvens": break_evens,
        "margem": margem,
        "premioLiq": premio_liq,
        "delta": g_delta if has_greeks else None,
        "gamma": g_gamma if has_greeks else None,
        "theta": g_theta if has_greeks else None,
        "vega": g_vega if has_greeks else None,
        "rho": g_rho if has_greeks else None,
    }
